### PR TITLE
test: add local deployments for custom component

### PR DESCRIPTION
**Pull request by dgggIoT**
**Issue #, if available:**

**Description of changes:**
This only serves for adding the basic flow and structure of local deployment of custom component. The custom component will be in next PR.
**Why is this change necessary:**

**How was this change tested:**

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: uat/testing-features/pom.xml [hunk 1]
```diff
@@ -150,6 +165,16 @@
...
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.7.0</version>
+                <configuration>
+                    <source>1.8</source>
+                    <target>1.8</target>
+                    <encoding>UTF-8</encoding>
+                </configuration>
+            </plugin>
```
Comment (saranyailla) [inline uat/testing-features/pom.xml:177]: You can revert this and also the lines 16 and 17.

File: uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                "--artifactDir " + artifactPath.toString(),
+                "--recipeDir " + recipePath.toString()));
+        if (StringUtils.isNotBlank(componentVersion)) {
+            commandArgs.add("--merge " + componentName + "=" + componentVersion);
+        }
+        String updateConfigArgs = getCliUpdateConfigArgs(componentName, configuration);
+        if (!updateConfigArgs.isEmpty()) {
+            commandArgs.add("--update-config '" + updateConfigArgs + "'");
+        }
+        LOGGER.info("The cli command: {}", commandArgs);
```
Comment (saranyailla) [inline uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java:134]: We can remove this log.

File: uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    /**
+     * implemented the step of installing a custom component with configuration.
+     *
+     * @param componentName         the name of the custom component
+     * @param configurationTable    the table which describes the configurations
+     * @throws InterruptedException InterruptedException could be throw out during the component deployment
+     * @throws IOException          IOException could be throw out during preparation of the CLI command
+     */
+    @When("I install the component {word} from local store with configuration")
```
Comment (MikeDombo) [inline uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java:92]: why is this here?

Local deployments should be implemented in OTF itself, this is not specific to Log Manager. Can we do this better?

**Review by saranyailla — DISMISSED** [review]

File: uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature [hunk 4]
```diff
@@ -9,6 +9,18 @@ Feature: Greengrass V2 LogManager
         And 5 temporary rotated log files for component aws.greengrass.Nucleus have been created
         And 5 temporary rotated log files for component UserComponentA have been created
 
+    @local
+    Scenario: LogManager-1-T0: install the local custom component
+        Given I create a Greengrass deployment with components
+            | aws.greengrass.Cli        | LATEST |
```
Comment (MikeDombo) [inline uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature:15]: OTF doesn't have the CLI by default?

File: uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature [hunk 5]
```diff
@@ -9,6 +9,18 @@ Feature: Greengrass V2 LogManager
...
+        And I deploy the Greengrass deployment configuration
+        Then the Greengrass deployment is COMPLETED on the device after 2 minutes
+        When I install the component logGenerator from local store with configuration
```
Comment (MikeDombo) [inline uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature:18]: I don't want to repeat the same problem that we have with our internal tests.

I want this to use MERGE/RESET format directly.


```
"""json
{
"MERGE": {"key": "value", "FileSize": "10MB"}
}
"""
```

File: uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java (hunk 3 above)
Comment (dgggIoT) [reply uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java:92]: OTF has local deployment but it does not take parameters. The LogManager has it here so then it can take some parameters for different scenarios, e.g. logging frequency, file size, rotation speed.

Comment (MikeDombo) [reply uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java:92]: OTF should have this logic.

Comment (dgggIoT) [reply uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java:92]: But for different component, it is hard to define the number of parameters. Will it be more component-owned thing, instead of general steps in OTF?  Here is the [link](https://github.com/aws-greengrass/aws-greengrass-testing/blob/1ce7ef461465fb11f701f2e1231caa817afcba04/aws-greengrass-testing-features/aws-greengrass-testing-features-api/src/main/java/com/aws/greengrass/testing/features/DeploymentSteps.java#L148) for what original OTF has:

File: uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature (hunk 4 above)
Comment (dgggIoT) [reply uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature:15]: OTF does not install CLI automatically.

File: uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java (hunk 3 above)
Comment (dgggIoT) [reply uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java:92]: This will be moved into OTF eventually. Now, we are trying what we want for local deployments and will provide feedback to OTF team.

Comment (MikeDombo) [reply uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java:92]: We're all the same team. Why do you add the appropriate code into OTF and then consume it here? We don't need to wait for them to do it, you've already done it; just add it into OTF with a pull request.

Comment (dgggIoT) [reply uat/testing-features/src/main/java/com/aws/greengrass/LocalDeploymentSteps.java]: The codes have been merged into OTF. The new commit is letting LogManager directly consume methods from OTF.

**Review by MikeDombo — APPROVED** [review]

**Review by saranyailla — APPROVED** [review]

File: uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature [hunk 6]
```diff
@@ -9,6 +9,24 @@ Feature: Greengrass V2 LogManager
...
+        Given I create a Greengrass deployment with components
+            | aws.greengrass.Cli        | LATEST |
+        And I deploy the Greengrass deployment configuration
+        Then I wait 120 seconds
+        Then the Greengrass deployment is COMPLETED on the device after 2 minutes
+        When I install the component logGenerator from local store with configuration
+        """
+        {
+            "MERGE": {
+                "FileSize": "10MB",
```
Comment (Nelsonochoam) [inline uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature:23]: Maybe we should split the unit and size so parsing becomes easier. Also speed might not be the appropriate wording in this context (speed = distance traveled / time elapsed) and we are going to use it to how frequent we write a log into a file. Maybe try



```suggestion
                "FileSize": "10",
                "FileSizeUnit": "MB",
                "LogWriteFrequencySeconds": 1
```

File: uat/testing-features/pom.xml [hunk 7]
```diff
@@ -103,6 +103,21 @@
                 <artifactId>maven-resources-plugin</artifactId>
                 <version>2.6</version>
                 <executions>
+                    <execution>
```
Comment (Nelsonochoam) [inline uat/testing-features/pom.xml:106]: It would be nice if OTF did this for us and we had a specific structure it could look for (just a thought for the future)

File: uat/custom-components/src/main/resources/recipes/logGenerator.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+
+---
+RecipeFormatVersion: '2020-01-25'
+ComponentName: logGenerator
+ComponentVersion: '0.0.0'
+ComponentDescription: Component to generate log files with random log messages.
+ComponentPublisher: AWS
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/resources/recipes/logGenerator.yaml:11]:
```suggestion
ComponentConfiguration:
     DefaultConfiguration:
         LogWriteFrequencySeconds: 1
```

File: uat/custom-components/src/main/resources/recipes/logGenerator.yaml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+Manifests:
+  - Artifacts:
+      - URI: classpath:/local-store/artifacts/custom-components.jar
+    Lifecycle:
+      Run: >-
+        java -Dlog.level=INFO -DcomponentName="logGenerator" -jar {artifacts:path}/custom-components.jar 
+        "{configuration:/FileSize}" "{configuration:/LogSpeed}"
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/resources/recipes/logGenerator.yaml:18]:
```suggestion
        "{configuration:/FileSize}" "configuration:/FileSizeUnit" "{configuration:/LogWriteFrequencySeconds}"
```

File: uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature (hunk 6 above)
Comment (dgggIoT) [reply uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature:23]: This is just temporary for testing if the current toy component can receive parameters. They should be well-defined once the custom component is developed.

File: uat/custom-components/src/main/resources/recipes/logGenerator.yaml (hunk 8 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/resources/recipes/logGenerator.yaml:11]: Why do we want to remove the publisher?

File: uat/custom-components/src/main/resources/recipes/logGenerator.yaml (hunk 9 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/resources/recipes/logGenerator.yaml:18]: This is just temporary for testing if the current toy component can receive parameters. They should be well-defined once the custom component is developed.

File: uat/custom-components/src/main/resources/recipes/logGenerator.yaml (hunk 8 above)
Comment (Nelsonochoam) [reply uat/custom-components/src/main/resources/recipes/logGenerator.yaml:11]: We don't that was a mistake

File: uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature (hunk 6 above)
Comment (Nelsonochoam) [reply uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature:23]: Why even add any value then?

Comment (dgggIoT) [reply uat/testing-features/src/main/resources/greengrass/features/log-manager-1.feature:23]: Good suggestion. I will apply them when developing custom component
